src/base/search/searchhandler.cpp

```diff
@@ -0,0 +1,199 @@
+    m_searchTimeout->stop();
+}
+
+// Slot called when QProcess is Finished
+// QProcess can be finished for 3 reasons:
+// Error | Stopped by user | Finished normally
+void SearchHandler::processFinished(int exitcode)
+{
+    m_searchTimeout->stop();
+
+    if (m_searchCancelled)
+        emit searchFinished(true);
+    else if ((m_searchProcess->exitStatus() == QProcess::NormalExit) && (exitcode == 0))
+        emit searchFinished(false);
+    else
+        emit searchFailed();
+}
+
+// search QProcess return output as soon as it gets new
+// stuff to read. We split it into lines and parse each
+// line to SearchResult calling parseSearchResult().
+void SearchHandler::readSearchOutput()
+{
+    QByteArray output = m_searchProcess->readAllStandardOutput();
+    output.replace("\r", "");
+    QList<QByteArray> lines = output.split('\n');
+    if (!m_searchResultLineTruncated.isEmpty())
+        lines.prepend(m_searchResultLineTruncated + lines.takeFirst());
+    m_searchResultLineTruncated = lines.takeLast().trimmed();
+
+    QList<SearchResult> searchResultList;
+    foreach (const QByteArray &line, lines) {
+        SearchResult searchResult;
+        if (parseSearchResult(QString::fromUtf8(line), searchResult))
+            searchResultList << searchResult;
+    }
+
+    if (!searchResultList.isEmpty()) {
+        m_results.append(searchResultList);
+        emit newSearchResults(searchResultList);
+    }
+}
+
+void SearchHandler::processFailed()
+{
+    if (!m_searchCancelled)
+        emit searchFailed();
+}
+
+// Parse one line of search results list
+// Line is in the following form:
+// file url | file name | file size | nb seeds | nb leechers | Search engine url
+bool SearchHandler::parseSearchResult(const QString &line, SearchResult &searchResult)
+{
+    const QStringList parts = line.split("|");
+    const int nbFields = parts.size();
+    if (nbFields < (NB_PLUGIN_COLUMNS - 1)) return false; // -1 because desc_link is optional
+
+    searchResult = SearchResult();
+    searchResult.fileUrl = parts.at(PL_DL_LINK).trimmed(); // download URL
+    searchResult.fileName = parts.at(PL_NAME).trimmed(); // Name
+    searchResult.fileSize = parts.at(PL_SIZE).trimmed().toLongLong(); // Size
+    bool ok = false;
+    searchResult.nbSeeders = parts.at(PL_SEEDS).trimmed().toLongLong(&ok); // Seeders
+    if (!ok || (searchResult.nbSeeders < 0))
+        searchResult.nbSeeders = -1;
+    searchResult.nbLeechers = parts.at(PL_LEECHS).trimmed().toLongLong(&ok); // Leechers
+    if (!ok || (searchResult.nbLeechers < 0))
+        searchResult.nbLeechers = -1;
+    searchResult.siteUrl = parts.at(PL_ENGINE_URL).trimmed(); // Search site URL
+    if (nbFields == NB_PLUGIN_COLUMNS)
+        searchResult.descrLink = parts.at(PL_DESC_LINK).trimmed(); // Description Link
+
+    return true;
+}
+
+SearchPluginManager *SearchHandler::manager() const
+{
+    return m_manager;
+}
+
+QList<SearchResult> SearchHandler::results() const
+{
+    return m_results;
```

Comment:
This function isn't really helpful in its current state. You can only retrieve the entire list, which means it can't provide incremental updates. There's a signal that sends only the new results, but I'd still need a queue in the web implementation

Reply:
>This function isn't really helpful in its current state.

If something isn't helpful for you it is not a reason to say "it sn't really helpful".
>You can only retrieve the entire list

You can retrieve the currently available results at any time. The all results become available when search job is finished.
> I'd still need a queue in the web implementation

I'm not sure you really need a Queue. However, I will be able to understand it only after seeing your implementation.

I'll try to explain it as a whole.
From a design point of view, base classes cannot strive to meet all the possible needs of all their "clients". They should logically, coherently and consistently implement some model from some subject area.
The SearchHandler tracks the some particular search job, accumulates the search results, and provides access to them. It implements **one-to-many** relationship (i.e. many observers can observe one search job). Its interface is enough to satisfy any of its client. But this does not mean that the client should not do anything else itself. In some cases, you may need some additional processing on the client side (I hope you understand the context in which I use the term "client").

Reply:
> If something isn't helpful for you it is not a reason to say "it sn't really helpful".

You don't use the function anywhere, so I assumed that's why you added it. I can't make use of it either.

> I'm not sure you really need a Queue. However, I will be able to understand it only after seeing your implementation.

The web client polls for new results at a fixed interval. I'll need to maintain a queue of explicit "new" results.

Reply:
>You don't use the function anywhere, so I assumed that's why you added it. I can't make use of it either.

This method is logical in the terms of the implemented model. It's not so expensive to worry much about it (I mean the accumulation of result at all). Besides, I can't know exactly which way you're gonna take. If in the end we find it too costly to have it unused, we will be able to remove it at any time.

>I'll need to maintain a queue of explicit "new" results.

Why don't you use "sync" way?

Reply:
> Why don't you use "sync" way?

I'm not familiar with sync. What is that?

Reply:
I just mean the way used in syncController.
In a nutshell, we compare current and previous data, and send only the difference.

Reply:
Anyway the main requirement is you must not implement the Search API in order to focus only on one specific use case (i.e. your client implementation). Its interface must be logical in itself.

Reply:
> In a nutshell, we compare current and previous data, and send only the difference.

I plan to support tabbed searching, which seems like it could be a lot of data to load into the session. With multiple searches returning 300 results/second, I don't know if this is outside the scope of session storage.

Reply:
Okay, let's see what it looks like. 
If no blocking comments occur here all of a sudden, I'll merge it tomorrow or the day after tomorrow. Then you can rebase your PR and we start to discuss it in details.